feat(derive+trusted-sync): online blob provider with fallback

merklefruit opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

Added a fallback online blob provider in the form of `OnlineBlobProviderWithFallback`.
It will first try to fetch blobs from the primary provider, and proceed with the fallback one if the primary fails.

I added the `BlobSidecarProvider` trait because blob archive APIs aren't required to implement all functions contained in `BeaconClient` (as is the case for https://github.com/base-org/blob-archiver). Furthermore, `genesis_time` and `slot_interval` are taken from the primary provider which is already an online beacon client.

Would something like a `BlobArchiver` struct implementation be useful? 
I'm not sure if that should be left to the caller or to the library.

**Tests**

Added unit tests to make sure that blob fetching is completed even without any sidecar loaded in the primary.

**Additional context**

I need some pointers on error handling and metrics, as I'm not sure of the standard used across the codebase.

**Metadata**

- Fixes #316

refcell commented on crates/derive/src/online/blob_provider.rs:
```diff
@@ -176,6 +176,168 @@ where
+        crate::timer!(START, PROVIDER_RESPONSE_TIME, &["blob_provider", "get_blobs"], timer);
+        match self.primary.get_blobs(block_ref, blob_hashes).await {
+            Ok(blobs) => Ok(blobs),
+            Err(_primary_err) => {
```
We should log the primary error here. If we're seeing much traffic routed to the fallback provider, we'd want to investigate the logs to see why the primary provider is throwing errors.

refcell commented on crates/derive/src/online/pipeline.rs:
```diff
@@ -18,7 +18,7 @@ pub type OnlinePipeline =
 pub type OnlineDataProvider = EthereumDataSource<
     AlloyChainProvider,
-    OnlineBlobProvider<OnlineBeaconClient, SimpleSlotDerivation>,
+    OnlineBlobProviderWithFallback<OnlineBeaconClient, OnlineBeaconClient, SimpleSlotDerivation>,
```
I don't really mind doing this, but can we make the fallback optional so that the current functionality is backwards compatible?

refcell commented on examples/trusted-sync/src/main.rs:
```diff
@@ -39,6 +39,7 @@ async fn sync(cli: cli::Cli) -> Result<()> {
     let l1_rpc_url = cli.l1_rpc_url()?;
     let l2_rpc_url = cli.l2_rpc_url()?;
     let beacon_url = cli.beacon_url()?;
+    let blob_archiver_url = cli.blob_archiver_url()?;
```
Even thought the argument is optional, this will cause the application to crash since the error is returned here.

We should introduce a builder pattern here IMO.
```
OnlineBlobProviderBuilder::new().with_beacon_client(beacon_client).with_fallback(Option<block_archiver_url>).build()?
```

This could return either the `OnlineBlobProvider` if no fallback is provided or `OnlineBlobProviderWithFallback` otherwise.

merklefruit replied:
Good point! Adding!

refcell reviewed (changes requested):
Request for backwards compatibility and a builder pattern

merklefruit commented:
@refcell Addressed the review. Just one point: `OnlineProviderBuilder::build()` now always returns `OnlineBlobProviderWithFallback` instead of two possible types depending if the fallback is defined. I can add that functionality with an extra typestate variable, however it would break the types for the convenience method `new_online_pipeline` used in the example.

In short, these are the two options:

1. (current implementation) always work with an `OnlineBlobProviderWithFallback`, which will act just like a regular `OnlineBlobProvider` in the case no fallback is added at build time. If a blob is not found, it will error with a clear message stating that blobs couldn't be retrieved and no fallback is configured.
2. break the types of `new_online_pipeline` and work with a generic `BlobProvider` trait in the example


I personally like (1) more, wdyt? Is there a better option? Open for feedback :)

refcell commented on examples/trusted-sync/src/main.rs:
```diff
@@ -71,9 +72,14 @@ async fn sync(cli: cli::Cli) -> Result<()> {
-    let beacon_client = OnlineBeaconClient::new_http(beacon_url);
-    let blob_provider =
-        OnlineBlobProvider::<_, SimpleSlotDerivation>::new(beacon_client, None, None);
+    let blob_provider = if let Some(blob_archiver_url) = blob_archiver_url {
+        OnlineBlobProviderBuilder::new()
+            .with_primary_beacon_client_url(beacon_url)
+            .with_fallback_beacon_client_url(blob_archiver_url)
+            .build()
+    } else {
+        OnlineBlobProviderBuilder::new().with_primary_beacon_client_url(beacon_url).build()
+    };
```
Can we just have `with_fallback_beacon_client_url` accept an Option here?

Then we can just do:
```rust
let blob_provider = OnlineBlobProviderBuilder::new().with_primary(beacon_url).with_fallback(cli.blob_archiver_url()).build()

```

refcell reviewed (approved):
Very nice!!
